lib/models/os.coffee

```diff
@@ -14,23 +14,223 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ###
 
+Promise = require('bluebird')
+reject = require('lodash/reject')
+once = require('lodash/once')
+semver = require('semver')
+
 errors = require('resin-errors')
-{ onlyIf } = require('../util')
+
+{ onlyIf, getImgMakerHelper, findCallback, notFoundResponse, deviceTypes: deviceTypesUtil } = require('../util')
 
 getOsModel = (deps, opts) ->
 	{ request } = deps
 	{ isBrowser, imageMakerUrl } = opts
 
+	imgMakerHelper = getImgMakerHelper(imageMakerUrl, request)
+
+	configModel = once -> require('./config')(deps, opts)
+
+	getDeviceTypes = once ->
+		configModel().getDeviceTypes()
+
+	isValidDeviceType = (deviceType) ->
+		getDeviceTypes()
+		.then (types) ->
+			!!deviceTypesUtil.findBySlug(types, deviceType)
+
+	getDownloadSize = imgMakerHelper.buildApiRequester
+		url: '/size_estimate'
+		withVersion: true
+		postProcess: ({ body }) -> body.size
+
+	getOsVersions = imgMakerHelper.buildApiRequester
+		buildUrl: ({ deviceType }) ->
+			"/image/#{deviceType}/versions"
+		postProcess: ({ body }) ->
+			{ versions, latest } = body
+			versions.sort(semver.rcompare)
+			recommended = reject(versions, semver.prerelease)?[0] || null
+			return {
+				versions
+				recommended
+				latest
+				default: recommended or latest
+			}
+
+	getMaxSatisfyingVersion = (versionOrRange, osVersions) ->
+		if versionOrRange in [ 'default', 'latest', 'recommended' ]
+			return osVersions[versionOrRange]
+		return semver.getMaxSatisfying(osVersions.versions, versionOrRange)
+
+	normalizeVersion = (v) ->
+		if not v
+			throw new Error("Invalid version: #{v}")
+		if v is 'latest'
+			return v
+		vNormalized = if v[0] is 'v' then v.substring(1) else v
+		if not semver.valid(vNormalized)
+			throw new Error("Invalid semver version: #{v}")
+		return vNormalized
+
 	exports = {}
 
 	###*
-	# @summary Get OS image last modified date
+	# @summary Get OS download size estimate
+	# @name getDownloadSize
+	# @public
+	# @function
+	# @memberof resin.models.os
+	# @description **Note!** Currently only the raw (uncompressed) size is reported.
+	#
+	# @param {String} deviceType - device type slug
+	# @fulfil {Number} - OS image download size, in bytes.
+	# @returns {Promise}
+	#
+	# @example
+	# resin.models.os.getDownloadSize('raspberry-pi').then(function(size) {
+	# 	console.log('The OS download size for raspberry-pi', size);
+	# });
+	#
+	# resin.models.os.getDownloadSize('raspberry-pi', function(error, size) {
+	# 	if (error) throw error;
+	# 	console.log('The OS download size for raspberry-pi', size);
+	# });
+	###
+	exports.getDownloadSize = (deviceType, callback) ->
+		callback = findCallback(arguments)
+
+		return isValidDeviceType(deviceType)
+			.then (isValid) ->
+				if not isValid
+					throw new errors.ResinRequestError('No such device type')
```

Comment:
This (and the other similar cases below) shouldn't be throwing `ResinRequestError`. That's for HTTP errors, it comes with status code and body properties, this isn't anything like that. We've got a `ResinInvalidDeviceType` error already - we should use that instead.